visualize-admin/visualization-tool
# feat: Add a way to retrieve dimensions from multiple cubes

This PR adapts the application logic to support retrieving dimensions from multiple cubes at once.

### GQL / SPARQL
- Consolidated several GQL queries together into `DataCubesComponents` that supports retrieving the components from several cubes at once
- Removed `Components` query (replaced by `DataCubesComponents`)
- Removed `ComponentsWithHierarchies` and `DimensionHierarchy` queries (hierarchies are now fetched by default in `DataCubesComponents`). As we generally use them in many different places across the app, it's more developer-friendly to have one source of truth (and also, it might be even more performant to have one query, instead of several different ones fired from different parts of the app). We can also see how [the new version](https://github.com/visualize-admin/visualization-tool/pull/1223) of `query-cube-hierarchy` library improves performance, as then it might be even less worrying, or we can bring back the old way of separating hierarchies fetching to the specific places that use them
- Removed `DimensionValues` query (no need to have a separate query to fetch dimension values, as we can just specify one dimensionIri in  `DataCubesComponents` to get the same result)
- Removed dimensions & measures from `SearchCubes` query
- We will now get all hierarchy levels, not limited to 6 as before (by using scalar instead of nested retrieval)
- To further improve performance and limit resolver chains, I've extracted the dimension types from GQL to TypeScript. It also means that we now explicitly parse them inside the resolver
- Removed `DimensionMetadataFragment` type, which means we can now better type `Dimension` and `Measure`, instead of mostly using the type that contains both

### Misc
- Fixed default empty filter selection when initializing a segment and hierarchy is not present

---
## Context
This PR is a first part of preparations for merging the cubes. For the technical concept, see [this Notion document](https://www.notion.so/interactivethings/Combine-two-cubes-in-the-same-chart-d5711b414a7747e79edfc3cfe17d2c9b).

- [x] `DataCubesComponents`
- [ ] `DataCubesMetadata`
- [ ] `DataCubesObservations`

## How to test
1. Click around in the application and make sure nothing is broken 💥 
2. Open [GQL playground](https://visualization-tool-git-feat-several-cubes-in-the-sa-cef96f-ixt1.vercel.app/api/graphql) and send e.g. the following query to see that all dimensions from both cubes were fetched.

**Query**
```graphql
query DataCubesComponents(
  $sourceType: String!
  $sourceUrl: String!
  $locale: String!
  $filters: [DataCubeFilter!]!
) {
  dataCubesComponents(
    sourceType: $sourceType
    sourceUrl: $sourceUrl
    locale: $locale
    filters: $filters
  )
}
```

**Variables**
```js
{
  "sourceType": "sparql",
  "sourceUrl": "https://lindas.admin.ch/query",
  "locale": "en",
  "filters": [{ "iri": "https://energy.ld.admin.ch/sfoe/bfe_ogd84_einmalverguetung_fuer_photovoltaikanlagen/9", "latest": true }, { "iri": "https://environment.ld.admin.ch/foen/ubd0104/6", "latest": true }]
}
```
